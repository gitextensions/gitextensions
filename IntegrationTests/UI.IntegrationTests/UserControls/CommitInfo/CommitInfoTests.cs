using System.Reflection;
using ApprovalTests;
using CommonTestUtils;
using CommonTestUtils.MEF;
using FluentAssertions;
using GitCommands;
using GitCommands.Git;
using GitExtensions.UITests.CommandsDialogs;
using GitUI;
using GitUIPluginInterfaces;
using Microsoft.VisualStudio.Composition;
using NSubstitute;
using NUnit.Framework;
using ResourceManager;

namespace GitExtensions.UITests.UserControls.CommitInfo
{
    [Apartment(ApartmentState.STA)]
    public class CommitInfoTests
    {
        // Created once for the fixture
        private ReferenceRepository _referenceRepository;

        // Created once for each test
        private MockExecutable _gitExecutable;
        private GitUICommands _commands;

        [SetUp]
        public void SetUp()
        {
            ReferenceRepository.ResetRepo(ref _referenceRepository);
            _commands = new GitUICommands(_referenceRepository.Module);

            // mock git executable
            _gitExecutable = new MockExecutable();
            typeof(GitModule).GetField("_gitExecutable", BindingFlags.Instance | BindingFlags.NonPublic)
                .SetValue(_commands.Module, _gitExecutable);
            GitCommandRunner cmdRunner = new(_gitExecutable, () => GitModule.SystemEncoding);
            typeof(GitModule).GetField("_gitCommandRunner", BindingFlags.Instance | BindingFlags.NonPublic)
                .SetValue(_commands.Module, cmdRunner);

            var composition = TestComposition.Empty
                .AddParts(typeof(MockLinkFactory));
            ExportProvider mefExportProvider = composition.ExportProviderFactory.CreateExportProvider();
            ManagedExtensibility.SetTestExportProvider(mefExportProvider);
        }

        [OneTimeTearDown]
        public void OneTimeTearDown()
        {
            _referenceRepository.Dispose();
        }

        [Test]
        public void GetSortedTags_should_throw_on_git_warning()
        {
            RunCommitInfoTest(commitInfo =>
            {
                _gitExecutable.StageOutput(@"for-each-ref --sort=""-taggerdate"" --format=""%(refname)"" refs/tags/",
                    "refs/heads/master\nwarning: message");

                ((Action)(() => commitInfo.GetTestAccessor().GetSortedTags())).Should().Throw<RefsWarningException>();

                return Task.CompletedTask;
            });
        }

        [Test]
        public void GetSortedTags_should_split_output_if_no_warning()
        {
            RunCommitInfoTest(commitInfo =>
            {
                _gitExecutable.StageOutput(@"for-each-ref --sort=""-taggerdate"" --format=""%(refname)"" refs/tags/",
                    "refs/remotes/origin/master\nrefs/heads/master\nrefs/heads/warning"); // does not contain "warning:"

                Dictionary<string, int> expected = new()
                {
                    ["refs/remotes/origin/master"] = 0,
                    ["refs/heads/master"] = 1,
                    ["refs/heads/warning"] = 2
                };

                var refs = commitInfo.GetTestAccessor().GetSortedTags();

                refs.Count.Should().Be(3);
                refs.Should().BeEquivalentTo(expected);

                return Task.CompletedTask;
            });
        }

        [Test]
        public void GetSortedTags_should_load_ref_different_in_case()
        {
            RunCommitInfoTest(commitInfo =>
            {
                _gitExecutable.StageOutput(@"for-each-ref --sort=""-taggerdate"" --format=""%(refname)"" refs/tags/",
                    "refs/remotes/origin/master\nrefs/heads/master\nrefs/remotes/origin/bugfix/YS-38651-test-twist-changes-r100-on-s375\nrefs/remotes/origin/bugfix/ys-38651-test-twist-changes-r100-on-s375"); // case sensitive duplicates

                Dictionary<string, int> expected = new()
                {
                    ["refs/remotes/origin/master"] = 0,
                    ["refs/heads/master"] = 1,
                    ["refs/remotes/origin/bugfix/YS-38651-test-twist-changes-r100-on-s375"] = 2,
                    ["refs/remotes/origin/bugfix/ys-38651-test-twist-changes-r100-on-s375"] = 3
                };

                var refs = commitInfo.GetTestAccessor().GetSortedTags();

                refs.Count.Should().Be(4);
                refs.Should().BeEquivalentTo(expected);

                return Task.CompletedTask;
            });
        }

        [Test]
        public void GetSortedTags_should_load_ref_with_extra_spaces()
        {
            RunCommitInfoTest(commitInfo =>
            {
                _gitExecutable.StageOutput(@"for-each-ref --sort=""-taggerdate"" --format=""%(refname)"" refs/tags/",
                    "refs/remotes/origin/master\nrefs/heads/master\nrefs/tags/v3.1\nrefs/tags/v3.1 \n refs/tags/v3.1"); // have leading and trailing spaces

                Dictionary<string, int> expected = new()
                {
                    ["refs/remotes/origin/master"] = 0,
                    ["refs/heads/master"] = 1,
                    ["refs/tags/v3.1"] = 2,
                    ["refs/tags/v3.1 "] = 3,
                    [" refs/tags/v3.1"] = 4
                };

                var refs = commitInfo.GetTestAccessor().GetSortedTags();

                refs.Count.Should().Be(5);
                refs.Should().BeEquivalentTo(expected);

                return Task.CompletedTask;
            });
        }

        [Test]
        public void GetSortedTags_should_remove_duplicate_refs()
        {
            RunCommitInfoTest(commitInfo =>
            {
                _gitExecutable.StageOutput(@"for-each-ref --sort=""-taggerdate"" --format=""%(refname)"" refs/tags/",
                    "refs/remotes/origin/master\nrefs/remotes/foo/duplicate\nrefs/remotes/foo/bar\nrefs/remotes/foo/duplicate\nrefs/remotes/foo/last"); // exact duplicates

                Dictionary<string, int> expected = new()
                {
                    ["refs/remotes/origin/master"] = 0,
                    ["refs/remotes/foo/duplicate"] = 1,
                    ["refs/remotes/foo/bar"] = 2,
                    ["refs/remotes/foo/last"] = 3,
                };

                var refs = commitInfo.GetTestAccessor().GetSortedTags();

                refs.Count.Should().Be(4);
                refs.Should().BeEquivalentTo(expected);

                return Task.CompletedTask;
            });
        }

        [Test]
        public void ReloadCommitInfo_should_render_links_correctly()
        {
            string hash = "a48da1aba59a65b2a7f0df7e3512817caf16819f";

            // Generate branches: branch01...branch15
            _gitExecutable.StageOutput($"branch --contains {hash}", string.Join('\n', Enumerable.Range(1, 15).Select(i => $"branch{i:00}")));

            // Generate tags: v1.0...v1.15
            _gitExecutable.StageOutput($"tag --contains {hash}", string.Join('\n', Enumerable.Range(0, 15).Select(i => $"v1.{i}")));

            _gitExecutable.StageOutput($"describe --tags --first-parent --abbrev=40 {hash}", "");

            var realCommitObjectId = ObjectId.Parse(hash);
            GitRevision revision = new(realCommitObjectId)
            {
                Author = "John Doe",
                AuthorUnixTime = DateTimeUtils.ToUnixTime(DateTime.Parse("2010-03-24 13:37:12")),
                AuthorEmail = "j.doe@some.email.dotcom",
                Subject = "fix: bugs",
                Body = "fix: bugs\r\n\r\nall bugs fixed"
            };

            RunCommitInfoTest(async (commitInfo) =>
            {
                commitInfo.SetRevisionWithChildren(revision, children: null);

                // Wait for pending operations so the Control is loaded completely before testing it
                await AsyncTestHelper.JoinPendingOperationsAsync(AsyncTestHelper.UnexpectedTimeout);

                Approvals.Verify(commitInfo.GetTestAccessor().RevisionInfo.Text);
            });
        }

        // Link start numbers are obtained manually.
        [TestCase("branch05", 183, "gitext://gotobranch/branch05")]
        [TestCase("v1.9", 745, "gitext://gototag/v1.9")]
        public void ReloadCommitInfo_should_extract_links_correctly(string refText, int linkStart, string expectedUri)
        {
            string hash = "a48da1aba59a65b2a7f0df7e3512817caf16819f";
            string hashInBody = "a48da1aba59a65b2a7f0df7e3512817caf16819a";
            string hashLink = $"gitext://gotocommit/{hashInBody}";

            // Generate branches: branch01...branch15
            _gitExecutable.StageOutput($"branch --contains {hash}", string.Join('\n', Enumerable.Range(1, 15).Select(i => $"branch{i:00}")));

            // Generate tags: v1.0...v1.15
            _gitExecutable.StageOutput($"tag --contains {hash}", string.Join('\n', Enumerable.Range(0, 15).Select(i => $"v1.{i}")));

            _gitExecutable.StageOutput($"describe --tags --first-parent --abbrev=40 {hash}", "");

            var realCommitObjectId = ObjectId.Parse(hash);
            GitRevision revision = new(realCommitObjectId)
            {
                Author = "John Doe",
                AuthorUnixTime = DateTimeUtils.ToUnixTime(DateTime.Parse("2010-03-24 13:37:12")),
                AuthorEmail = "j.doe@some.email.dotcom",
                Subject = "fix: bugs",
                Body = $"fix: bugs\r\n\r\nall bugs from {hashInBody} fixed"
            };

            RunCommitInfoTest(async (commitInfo) =>
            {
                if (ManagedExtensibility.GetExport<ILinkFactory>().Value is not MockLinkFactory mockLinkFactory)
                {
                    Assert.Fail($"Unexpected {typeof(ILinkFactory)} implementation.");
                    return;
                }

                object commandClickedSender = null;
                commitInfo.CommandClicked += (s, e) => commandClickedSender = s;
                commitInfo.SetRevisionWithChildren(revision, children: null);

                // Wait for pending operations so the Control is loaded completely before testing it
                await AsyncTestHelper.JoinPendingOperationsAsync(AsyncTestHelper.UnexpectedTimeout);

                var ta = commitInfo.GetTestAccessor();

                // simulate a click on refText link
                ta.LinkClicked(ta.RevisionInfo, new(refText, linkStart, linkLength: refText.Length));
                mockLinkFactory.LastExecutedLinkUri.Should().Be(expectedUri);

                // simulate a click on hash link
                ta.LinkClicked(ta.CommitMessage, new(hashInBody, linkStart: 25, linkLength: hashInBody.Length));
                mockLinkFactory.LastExecutedLinkUri.Should().Be(hashLink);
                commandClickedSender.Should().Be(ta.CommitMessage);
            });
        }

        [Test]
        public void ReloadCommitInfo_should_handle_ShowAll_branches_correctly()
        {
            string hash = "a48da1aba59a65b2a7f0df7e3512817caf16819f";

            // Generate branches: branch01...branch15
            _gitExecutable.StageOutput($"branch --contains {hash}", string.Join('\n', Enumerable.Range(1, 15).Select(i => $"branch{i:00}")));

            // Generate tags: v1.0...v1.15
            _gitExecutable.StageOutput($"tag --contains {hash}", string.Join('\n', Enumerable.Range(0, 15).Select(i => $"v1.{i}")));

            _gitExecutable.StageOutput($"describe --tags --first-parent --abbrev=40 {hash}", "");

            var realCommitObjectId = ObjectId.Parse(hash);
            GitRevision revision = new(realCommitObjectId)
            {
                Author = "John Doe",
                AuthorUnixTime = DateTimeUtils.ToUnixTime(DateTime.Parse("2010-03-24 13:37:12")),
                AuthorEmail = "j.doe@some.email.dotcom",
                Subject = "fix: bugs",
                Body = "fix: bugs\r\n\r\nall bugs fixed"
            };

            RunCommitInfoTest(async (commitInfo) =>
            {
                if (ManagedExtensibility.GetExport<ILinkFactory>().Value is not MockLinkFactory mockLinkFactory)
                {
                    Assert.Fail($"Unexpected {typeof(ILinkFactory)} implementation.");
                    return;
                }

                commitInfo.SetRevisionWithChildren(revision, children: null);

                // Wait for pending operations so the Control is loaded completely before testing it
                await AsyncTestHelper.JoinPendingOperationsAsync(AsyncTestHelper.UnexpectedTimeout);

                // simulate a click on refText link
                var ta = commitInfo.GetTestAccessor();
                ta.LinkClicked(ta.RevisionInfo, new("not important", linkStart: 423, linkLength: 0));
                mockLinkFactory.LastExecutedLinkUri.Should().Be("gitext://showall/branches");

                Approvals.Verify(commitInfo.GetTestAccessor().RevisionInfo.Text);
            });
        }

        [Test]
        public void ReloadCommitInfo_should_handle_ShowAll_tags_correctly()
        {
            string hash = "a48da1aba59a65b2a7f0df7e3512817caf16819f";

            // Generate branches: branch01...branch15
            _gitExecutable.StageOutput($"branch --contains {hash}", string.Join('\n', Enumerable.Range(1, 15).Select(i => $"branch{i:00}")));

            // Generate tags: v1.0...v1.15
            _gitExecutable.StageOutput($"tag --contains {hash}", string.Join('\n', Enumerable.Range(0, 15).Select(i => $"v1.{i}")));

            _gitExecutable.StageOutput($"describe --tags --first-parent --abbrev=40 {hash}", "");

            var realCommitObjectId = ObjectId.Parse(hash);
            GitRevision revision = new(realCommitObjectId)
            {
                Author = "John Doe",
                AuthorUnixTime = DateTimeUtils.ToUnixTime(DateTime.Parse("2010-03-24 13:37:12")),
                AuthorEmail = "j.doe@some.email.dotcom",
                Subject = "fix: bugs",
                Body = "fix: bugs\r\n\r\nall bugs fixed"
            };

            RunCommitInfoTest(async (commitInfo) =>
            {
                if (ManagedExtensibility.GetExport<ILinkFactory>().Value is not MockLinkFactory mockLinkFactory)
                {
                    Assert.Fail($"Unexpected {typeof(ILinkFactory)} implementation.");
                    return;
                }

                commitInfo.SetRevisionWithChildren(revision, children: null);

                // Wait for pending operations so the Control is loaded completely before testing it
                await AsyncTestHelper.JoinPendingOperationsAsync(AsyncTestHelper.UnexpectedTimeout);

                // simulate a click on refText link
                var ta = commitInfo.GetTestAccessor();
                ta.LinkClicked(ta.RevisionInfo, new("not important", linkStart: 774, linkLength: 0));
                mockLinkFactory.LastExecutedLinkUri.Should().Be("gitext://showall/tags");

                Approvals.Verify(commitInfo.GetTestAccessor().RevisionInfo.Text);
            });
        }

        private void RunCommitInfoTest(Func<GitUI.CommitInfo.CommitInfo, Task> runTestAsync)
        {
            UITest.RunControl(
                createControl: form =>
                {
                    var uiCommandsSource = Substitute.For<IGitUICommandsSource>();
                    uiCommandsSource.UICommands.Returns(x => _commands);

                    // the following assignment of CommitInfo.UICommandsSource will already call this command
                    _gitExecutable.StageOutput(@"for-each-ref --sort=""-taggerdate"" --format=""%(refname)"" refs/tags/", "");

                    form.Size = new(600, 480);

                    return new GitUI.CommitInfo.CommitInfo
                    {
                        Dock = DockStyle.Fill,
                        Parent = form,
                        ShowBranchesAsLinks = true,
                        UICommandsSource = uiCommandsSource,
                    };
                },
                runTestAsync: async commitInfo =>
                {
                    // Wait for pending operations so the Control is loaded completely before testing it
                    await AsyncTestHelper.JoinPendingOperationsAsync(AsyncTestHelper.UnexpectedTimeout);

                    await runTestAsync(commitInfo);
                });
        }
    }
}
