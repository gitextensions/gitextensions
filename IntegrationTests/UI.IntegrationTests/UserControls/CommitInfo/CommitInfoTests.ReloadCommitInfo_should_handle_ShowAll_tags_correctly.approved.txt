{\rtf1\ansi\ansicpg1251\deff0\nouicompat\deflang1049{\fonttbl{\f0\fnil\fcharset204 Segoe UI;}}
{\colortbl ;\red0\green0\blue255;}
{\*\generator Riched20 10.0.22000}\viewkind4\uc1 
\pard\f0\fs18 Contained in branches:\par

\pard branch01\v |||gitext://gotobranch/branch01\v0\par
branch02\v |||gitext://gotobranch/branch02\v0\par
branch03\v |||gitext://gotobranch/branch03\v0\par
branch04\v |||gitext://gotobranch/branch04\v0\par
branch05\v |||gitext://gotobranch/branch05\v0\par
branch06\v |||gitext://gotobranch/branch06\v0\par
branch07\v |||gitext://gotobranch/branch07\v0\par
branch08\v |||gitext://gotobranch/branch08\v0\par
branch09\v |||gitext://gotobranch/branch09\v0\par
branch10\v |||gitext://gotobranch/branch10\v0\par
[ Show all ]\v |||gitext://showall/branches\v0\par
\par
Contained in tags:\par
v1.0\v |||gitext://gototag/v1.0\v0\par
v1.1\v |||gitext://gototag/v1.1\v0\par
v1.2\v |||gitext://gototag/v1.2\v0\par
v1.3\v |||gitext://gototag/v1.3\v0\par
v1.4\v |||gitext://gototag/v1.4\v0\par
v1.5\v |||gitext://gototag/v1.5\v0\par
v1.6\v |||gitext://gototag/v1.6\v0\par
v1.7\v |||gitext://gototag/v1.7\v0\par
v1.8\v |||gitext://gototag/v1.8\v0\par
v1.9\v |||gitext://gototag/v1.9\v0\par
v1.10\v |||gitext://gototag/v1.10\v0\par
v1.11\v |||gitext://gototag/v1.11\v0\par
v1.12\v |||gitext://gototag/v1.12\v0\par
v1.13\v |||gitext://gototag/v1.13\v0\par
v1.14\v |||gitext://gototag/v1.14\v0\par
\par
Derives from no tag\par
}
